compiler/rustc_next_trait_solver/src/solve/assembly/structural_traits.rs

```diff
@@ -871,72 +866,126 @@ where
             .extend(cx.item_bounds(associated_type_def_id).iter_instantiated(cx, trait_ref.args));
     }
 
-    let mut replace_projection_with = HashMap::default();
+    let mut replace_projection_with: HashMap<_, Vec<_>> = HashMap::default();
     for bound in object_bounds.iter() {
         if let ty::ExistentialPredicate::Projection(proj) = bound.skip_binder() {
+            // FIXME: We *probably* should replace this with a dummy placeholder,
+            // b/c don't want to replace literal instances of this dyn type that
+            // show up in the bounds, but just ones that come from substituting
+            // `Self` with the dyn type.
             let proj = proj.with_self_ty(cx, trait_ref.self_ty());
-            let old_ty = replace_projection_with.insert(proj.def_id(), bound.rebind(proj));
-            assert_eq!(
-                old_ty,
-                None,
-                "{:?} has two generic parameters: {:?} and {:?}",
-                proj.projection_term,
-                proj.term,
-                old_ty.unwrap()
-            );
+            replace_projection_with.entry(proj.def_id()).or_default().push(bound.rebind(proj));
         }
     }
 
-    let mut folder =
-        ReplaceProjectionWith { ecx, param_env, mapping: replace_projection_with, nested: vec![] };
-    let folded_requirements = requirements.fold_with(&mut folder);
+    let mut folder = ReplaceProjectionWith {
+        ecx,
+        param_env,
+        self_ty: trait_ref.self_ty(),
+        mapping: replace_projection_with,
+        nested: vec![],
+        is_ambiguous: false,
+    };
+
+    if folder.is_ambiguous {
+        return Err(Ambiguous);
+    }
 
-    folder
+    let requirements = requirements.fold_with(&mut folder);
```

Comment:
:thinking: 

i feel like the `if`  should be after the `fold_with` call :sweat_smile: 

is it possible to add a test where this ambiguity is actually relevant? I assume that ignoring `is_ambiguous` means we return the unnormalized alias and treat is as rigid here, encountering the solver cycle this folder wants to avoid :thinking:

Reply:
that is so funny, yeah it should happen before. my bad.

I'll try to think of a test where this matters -- perhaps something like `dyn Trait<?A, ?B>` where `Trait<A, B>: Sup<A, Assoc = A> + Sup<B, Assoc = B>` where we use need to replace `<Self as Sup<A>>::Assoc`, but then inference later constrains `?A` and `?B`.